VM Exec

tomponline opened the pull request:
Adds support for `lxc exec` for VMs via `lxd-agent`.

This is a WIP.

lxc-jenkins commented:
Testsuite passed

lxc-jenkins commented:
Testsuite passed

lxc-jenkins commented:
Testsuite passed

lxc-jenkins commented:
Testsuite passed

tomponline commented on client/lxd.go:
```diff
@@ -372,7 +372,7 @@ func (r *ProtocolLXD) rawWebsocket(url string) (*websocket.Conn, error) {
 
 	// Log the data
-	logger.Debugf("Connected to the websocket")
+	logger.Debugf("Connected to the websocket: %v", url)
```
@stgraber I noticed the url has a `secret` param on it, is this bad to be logging?

stgraber replied:
No, that's fine, it's a one time token effectively

lxc-jenkins commented:
Testsuite passed

lxc-jenkins commented:
Testsuite passed

lxc-jenkins commented:
Testsuite passed

brauner commented on lxd-agent/exec.go:
```diff
@@ -414,12 +425,25 @@ func (s *execWs) Do(op *operations.Operation) error {
 	cmd.Stdout = stdout
 	cmd.Stderr = stderr
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Credential: &syscall.Credential{
```
This should be taken from `https://godoc.org/golang.org/x/sys/unix` not from `syscall`, no?

tomponline replied:
Ah yes "Deprecated: this package is locked down. Callers should use the corresponding package in the golang.org/x/sys repository instead."

stgraber replied:
`exec.Command` depends on `syscall.SysProcAttr` directly, we don't get to use unix for this one

brauner replied:
_facepalm_

stgraber commented on lxd/vm_qemu.go:
```diff
@@ -2008,6 +2009,7 @@ func (vm *vmQemu) FilePush(fileType string, srcPath string, dstPath string, uid
 	if err != nil {
 		return err
 	}
+	agent.Disconnect()
```
Shouldn't that be a defer?

tomponline replied:
@stgraber indeed it should, although this didnt show up in my testing as FilePush doesnt actually wait on the channel that Disconnect() closes so it works fine, but I added what should have been defer agent.Disconnect() to all uses of agent for completeness. Fixed.

tomponline commented:
@stgraber I've updated this PR to lay the groundwork for handling signals sent to the `lxc` cli and forwarding them onto the `lxd-agent` cmd through LXD and vsock.

In order to do this I'm modified the `instance.Exec()` signature (and both `containerLXC` and `vmQemu` implementations) to always return after the command has been started (i.e removes the `wait` argument and always expects the caller to so this). It has also been modified to not return an `os.Cmd` but rather a new interface type called `instance.Cmd`.

The `instance.Cmd` type allows me to provide 2 implementations for, one for local commands run in containers via LXD/liblxc and one for remote commands run by `lxd-agent` from LXD via a websocket.

It exposes the following functions:

- PID()
- Wait()
- Signal()

This allows the existing users of the `Exec()` functions to do everything these used to be able to do with `os.Cmd` but in a type-agnostic way.

I've also noticed that there was a vsock socket leak if LXD was closed down whilst a command was active. I've partially fixed this by implementing a very basic no-op `Control` handler that was previously missing in the setup of `lxdClient.InstanceExecArgs` in `vmQemu.Exec()`. This has also fixed a go routine leak.

tomponline commented:
@stgraber I also noticed that although using `setsid` allows `ctrl+c` to work when executing a shell, such as `lxc exec v1 bash`, I've noticed that this does not work for running programs directly. For example running `top` means you cannot exit using ctrl+c.

I believe this is because the signal itself is received by the `lxc` cli and relayed out-of-band, so this should be handled by the signal handler work I mentioned above.

brauner commented:
> @stgraber I also noticed that although using `setsid` allows `ctrl+c` to work when executing a shell, such as `lxc exec v1 bash`, I've noticed that this does not work for running programs directly. For example running `top` means you cannot exit using ctrl+c.
> 
> I believe this is because the signal itself is received by the `lxc` cli and relayed out-of-band, so this should be handled by the signal handler work I mentioned above.

The signal needs to be handled and forwarded in both cases. See`lxc/exec_unix.go`.

stgraber commented:
```
/lxc-ci/build/tmp.4pu9AlTzH1/go/src/github.com/lxc/lxd/lxd/container_exec.go:522:9: ineffectual assignment to err
```

tomponline commented:
@stgraber thats fixed now.

lxc-jenkins commented:
Testsuite passed

stgraber commented:
@tomponline cool, so that's moved the logic around and still need Control to be properly hooked up right?

tomponline commented:
@stgraber Yep, that is correct, the Exec is still working as it was, but with the new interface implemented.

- ctrl+c works if you run `lxc exec v1 bash` and then execute another program from the remote shell
- ctrl+c (and other signals) don't work if executing a remote program directly (such as `lxc exec v1 top`) as the signals sent to the local `lxc` cli are not yet forwarded to the lxd-agent and remote program.

Will tackle that on Monday.

tomponline commented:
@brauner @stgraber I would appreciate your advice on how best to provide "ctrl+c" semantics to `lxc exec` for VMs.

I've just pushed up signal forwarding changes to this PR, which means if I do (for example) `lxc exec v1 top` and then in another shell do `killall -s SIGUS1 lxc` I can see that the signal is forwarded by lxc to lxd and onto the lxd-agent, finally being passed to the remote `top` process in the VM.

However when running `lxc exec v1 top` and pressing `ctrl+c` still doesn't allow the program to exit.

My thinking behind is that because both the local lxc terminal and the remote lxd -> lxd-agent PTY are in raw mode, the `ctrl+c` combination is being forwarded literally, and that the remote command doesn understand ctrl+c as literal input to mean exit (it expects the shell to convert this to a signal).

This would explain why running `lxc exec v1 bash` and then starting top from inside the shell does allow `ctrl+c` to exit the top command.

My question is do you have any suggestions as to where the `ctrl+c` sequence should be converted into a signal, and how to we avoid interfering with programs that install their own signal handlers? Thanks

lxc-jenkins commented:
Testsuite passed

lxc-jenkins commented:
Testsuite passed

lxc-jenkins commented:
Testsuite passed

lxc-jenkins commented:
Testsuite passed
